Putting HashMap in a heap block hurts memory use, speed, code size (WebCore databases)

darinadler opened the pull request:
#### d73b4b3064358261cd1f0ee3585e2b103db6b8db
<pre>
Putting HashMap in a heap block hurts memory use, speed, code size (WebCore databases)
<a href="https://bugs.webkit.org/show_bug.cgi?id=295861">https://bugs.webkit.org/show_bug.cgi?id=295861</a>
<a href="https://rdar.apple.com/155734983">rdar://155734983</a>

Reviewed by Sihui Liu and Sam Weinig.

HashMap and related class templates already are pointers to a hash table in
production builds, so there is no reason to wrap them in heap block with
unique_ptr and add another level of indirection. Fix this in a few places
in WebCore database modules.

* Source/WebCore/Modules/indexeddb/server/MemoryBackingStoreTransaction.h:
Revise objectStoreCleared to take a map instead of a pointer to a map.

* Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
(WebCore::IDBServer::MemoryObjectStore::forEachRecord): Update since
m_keyValueStore is now a map rather than a pointer to a map.
(WebCore::IDBServer::MemoryObjectStore::containsRecord): Ditto.
(WebCore::IDBServer::MemoryObjectStore::clear): Ditto.
(WebCore::IDBServer::MemoryObjectStore::deleteRecord): Ditto.
(WebCore::IDBServer::MemoryObjectStore::addRecordWithoutUpdatingIndexes): Ditto.
(WebCore::IDBServer::MemoryObjectStore::addRecord): Ditto.
(WebCore::IDBServer::MemoryObjectStore::countForKeyRange const): Ditto.
(WebCore::IDBServer::MemoryObjectStore::valueForKey const): Ditto.
(WebCore::IDBServer::MemoryObjectStore::valueForKeyRange const): Ditto.
(WebCore::IDBServer::MemoryObjectStore::lowestKeyWithRecordInRange const): Ditto.
(WebCore::IDBServer::MemoryObjectStore::registerIndex): Ditto.
(WebCore::IDBServer::MemoryObjectStore::renameIndex): Ditto.
* Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.h: Ditto.

* Source/WebCore/Modules/webdatabase/DatabaseTracker.cpp:
(WebCore::DatabaseTracker::openDatabases): Update since m_openDatabaseMap is now a
map containing maps rather than a pointer to a map containing pointers to maps.
(WebCore::DatabaseTracker::addOpenDatabase): Ditto.
(WebCore::DatabaseTracker::removeOpenDatabase): Ditto.
(WebCore::DatabaseTracker::deleteDatabaseFile): Ditto.
* Source/WebCore/Modules/webdatabase/DatabaseTracker.h: Ditto.

Canonical link: <a href="https://commits.webkit.org/297507@main">https://commits.webkit.org/297507@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c017a5a5efe44f1bbf951dac76e17c366b3acff7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111374 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31040 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21505 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117406 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61642 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113336 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31721 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39622 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84647 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114321 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25329 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100263 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65094 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24672 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18400 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61226 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94714 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18467 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120596 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38423 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28553 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93571 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/fractional-translation-from-position.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38799 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96534 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93395 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23929 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38502 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16274 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34439 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38312 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43789 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37977 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41310 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39679 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
```diff
@@ -313,12 +299,10 @@ void MemoryObjectStore::addRecordWithoutUpdatingIndexes(MemoryBackingStoreTransa
 
-    if (!m_keyValueStore) {
-        m_keyValueStore = makeUnique<KeyValueMap>();
+    if (m_orderedKeys)
```
This should be `!m_orderedKeys` (tests might fail for this)

darinadler replied:
I think you spotted it. Thank you!

weinig commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
```diff
@@ -230,25 +227,17 @@ void MemoryObjectStore::deleteAllIndexes(MemoryBackingStoreTransaction& transact
-        m_orderedKeys = nullptr;
-    }
+    for (auto& [key, value] : std::exchange(m_keyValueStore, { }))
+        m_transactionModifiedRecords.add(key, value);
```
We should really have a way to bulk add entries to a HashMap as it could potentially avoid multiple allocations by reserving the appropriate final capacity!

darinadler replied:
I think we do have a capacity reserving idiom we could use here. But something that's _both_ efficient and idiomatic is better.

weinig commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
```diff
@@ -263,19 +252,16 @@ void MemoryObjectStore::deleteRecord(const IDBKeyData& key)
-        m_transactionModifiedRecords.set(key, iterator->value);
-
-    m_keyValueStore->remove(iterator);
+    m_keyValueStore.remove(iterator);
```
Would this be more efficient if instead of doing a `find` + `remove` it was just a `take`?

darinadler replied:
I don't think `take` is more efficient than `find`/`remove`. I admit, though, that it is easier to use correctly since it doesn't involve iterator lifetime rules. Not sure how easy this would be to refactor to use `take`. Maybe very easy!

weinig commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
```diff
@@ -332,31 +316,26 @@ IDBError MemoryObjectStore::addRecord(MemoryBackingStoreTransaction& transaction
     }
 
-    auto mapResult = m_keyValueStore->set(keyData, value.data());
+    auto mapResult = m_keyValueStore.add(keyData, value.data());
```
Just to double check, replacing `set` with `add` here is ok because we assert that m_keyValueStore does not already have `keyData` above?

darinadler replied:
That's right. And when we can do either we prefer `add`, because `set` is just implemented as `add` followed by assignment if the `add` doesn't make a new entry.

weinig commented on Source/WebCore/Modules/webdatabase/DatabaseTracker.cpp:
```diff
@@ -524,12 +524,10 @@ Vector<Ref<Database>> DatabaseTracker::openDatabases()
+        for (auto& nameMap : m_openDatabaseMap.values()) {
+            for (auto& set : nameMap.values()) {
+                for (auto& database : set)
+                    openDatabases.append(*database);
```
Could possibly use `openDatabases.appendRange(set.begin(), set.end())` (sidenote, we should add an overload that takes a generic container/std::range)?

darinadler replied:
I don't think that will convert raw pointers into `Ref`, which I believe is what this does.

weinig commented on Source/WebCore/Modules/webdatabase/DatabaseTracker.cpp:
```diff
@@ -540,65 +538,43 @@ void DatabaseTracker::addOpenDatabase(Database& database)
+    auto name = database.stringIdentifierIsolatedCopy();
+    m_openDatabaseMap.add(database.securityOrigin().isolatedCopy(), DatabaseNameMap { }).iterator->value
+        .add(name, DatabaseSet { }).iterator->value
+        .add(&database);
```
Nice.

Do you dare move all three adds?

```c++
    m_openDatabaseMap
        .add(database.securityOrigin().isolatedCopy(), DatabaseNameMap { }).iterator->value
        .add(name, DatabaseSet { }).iterator->value
        .add(&database);
```

darinadler replied:
I wouldn't have thought of that but I will do it.

weinig reviewed (approved):
Any idea why these were in unique_ptrs to begin with?

darinadler committed:
Putting HashMap in a heap block hurts memory use, speed, code size (WebCore databases)
https://bugs.webkit.org/show_bug.cgi?id=295861
rdar://155734983

Reviewed by Sihui Liu and Sam Weinig.

HashMap and related class templates already are pointers to a hash table in
production builds, so there is no reason to wrap them in heap block with
unique_ptr and add another level of indirection. Fix this in a few places
in WebCore database modules.

* Source/WebCore/Modules/indexeddb/server/MemoryBackingStoreTransaction.h:
Revise objectStoreCleared to take a map instead of a pointer to a map.

* Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
(WebCore::IDBServer::MemoryObjectStore::forEachRecord): Update since
m_keyValueStore is now a map rather than a pointer to a map.
(WebCore::IDBServer::MemoryObjectStore::containsRecord): Ditto.
(WebCore::IDBServer::MemoryObjectStore::clear): Ditto.
(WebCore::IDBServer::MemoryObjectStore::deleteRecord): Ditto.
(WebCore::IDBServer::MemoryObjectStore::addRecordWithoutUpdatingIndexes): Ditto.
(WebCore::IDBServer::MemoryObjectStore::addRecord): Ditto.
(WebCore::IDBServer::MemoryObjectStore::countForKeyRange const): Ditto.
(WebCore::IDBServer::MemoryObjectStore::valueForKey const): Ditto.
(WebCore::IDBServer::MemoryObjectStore::valueForKeyRange const): Ditto.
(WebCore::IDBServer::MemoryObjectStore::lowestKeyWithRecordInRange const): Ditto.
(WebCore::IDBServer::MemoryObjectStore::registerIndex): Ditto.
(WebCore::IDBServer::MemoryObjectStore::renameIndex): Ditto.
* Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.h: Ditto.

* Source/WebCore/Modules/webdatabase/DatabaseTracker.cpp:
(WebCore::DatabaseTracker::openDatabases): Update since m_openDatabaseMap is now a
map containing maps rather than a pointer to a map containing pointers to maps.
(WebCore::DatabaseTracker::addOpenDatabase): Ditto.
(WebCore::DatabaseTracker::removeOpenDatabase): Ditto.
(WebCore::DatabaseTracker::deleteDatabaseFile): Ditto.
* Source/WebCore/Modules/webdatabase/DatabaseTracker.h: Ditto.

Canonical link: https://commits.webkit.org/297507@main
